fineract-provider/src/main/java/org/apache/fineract/infrastructure/crypt/utils/RSAEncryptionUtils.java

```diff
@@ -0,0 +1,110 @@
+import org.apache.fineract.infrastructure.core.service.DateUtils;
+import org.apache.fineract.infrastructure.crypt.domain.EncryptionKeys;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Component;
+
+/**
+ * @author manoj
+ */
+@Component
+public class RSAEncryptionUtils {
+    private final static Logger LOG = LoggerFactory.getLogger(RSAEncryptionUtils.class);
+
+    public EncryptionKeys generateKeys(){
+        KeyPairGenerator keyPairGenerator;
+        try {
+            keyPairGenerator = KeyPairGenerator.getInstance("RSA");
+
+            KeyPair keyPair = keyPairGenerator.generateKeyPair();
+            PublicKey publicKey = keyPair.getPublic();
+            PrivateKey privateKey = keyPair.getPrivate();
+            byte[] publicKeyAsBytes = publicKey.getEncoded();
+            byte[] privateKeyAsBytes = privateKey.getEncoded();
+
+            return new EncryptionKeys(privateKeyAsBytes, publicKeyAsBytes, DateUtils.getLocalDateTimeOfTenant(), generateVersionString());
+
+        } catch (NoSuchAlgorithmException e) {
+           LOG.error("Error occurred", e);
+        }
+        throw new GeneralPlatformDomainRuleException("error.msg.error.in.generating.keys", "Error in generating keys");
+    }
+
+    public String decryptUsingRSA(String encryptedText, final byte[] privateKeyAsBytes, boolean isBase64Encoded) {
+        String decryptedString = null;
+        try {
+            KeyFactory kf = KeyFactory.getInstance("RSA");
+            PrivateKey privateKey = kf.generatePrivate(new PKCS8EncodedKeySpec(privateKeyAsBytes));
+            Cipher decrypt = Cipher.getInstance("RSA");
+            decrypt.init(Cipher.DECRYPT_MODE, privateKey);
+            if(isBase64Encoded) {
+                encryptedText = Hex.encodeHexString(encryptedText.getBytes( UTF_8));
+                byte[] decryptByteFromUi = decrypt.doFinal(DatatypeConverter.parseHexBinary(encryptedText));
+                decryptedString =  new String(decryptByteFromUi,  UTF_8);
+            } else {
+                byte[] decryptByteFromUi = decrypt.doFinal(encryptedText.getBytes( UTF_8));
+                decryptedString =  new String(decryptByteFromUi, UTF_8);
+            }
+            return decryptedString;
+        }catch (NoSuchAlgorithmException | InvalidKeySpecException | NoSuchPaddingException | InvalidKeyException |
+                IllegalBlockSizeException | BadPaddingException e) {
+            LOG.error("Error occurred", e);
+        }
+        throw new GeneralPlatformDomainRuleException("error.msg.error.in.decryption.keys", "Error in decryption");
+    }
+
+
+
+
+
+    private String generateVersionString() {
+        final char[] buf = new char[15];
+        final String alphanum = "ABCDEFGHIJKLMNOPQRSTUVWXYZ" + "0123456789" ;
+        final char[] symbols = alphanum.toCharArray();
+        final Random random= new SecureRandom();
```

Comment:
make this `SecureRandom` a private final field of this Component class instead of a variable in this method

Reply:
done!